File: ddd-application/src/main/java/victor/training/ddd/agile/application/service/SprintMetricsService.java
package victor.training.ddd.agile.application.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import victor.training.ddd.agile.application.dto.SprintMetrics;
import victor.training.ddd.agile.domain.model.BacklogItem;
import victor.training.ddd.agile.domain.model.Sprint;
import victor.training.ddd.agile.domain.model.Sprint.Status;
import victor.training.ddd.agile.domain.repo.SprintRepo;

import java.util.List;
import java.util.stream.Collectors;

@RequiredArgsConstructor
@Service
public class SprintMetricsService {
   private final SprintRepo sprintRepo;

   public SprintMetrics computeMetrics(long id) {
      Sprint sprint = sprintRepo.findOneById(id);
      if (sprint.getStatus() != Status.FINISHED) {
         throw new IllegalStateException();
      }
      List<BacklogItem> doneItems = sprint.getItems().stream()
          .filter(item -> item.getStatus() == BacklogItem.Status.DONE)
          .collect(Collectors.toList());
      SprintMetrics dto = new SprintMetrics();
      dto.consumedHours = sprint.getItems().stream().mapToInt(BacklogItem::getHoursConsumed).sum();
      dto.calendarDays = sprint.getStart().until(sprint.getEnd()).getDays();
      dto.doneFP = doneItems.stream().mapToInt(BacklogItem::getFpEstimation).sum();
      dto.fpVelocity = 1.0 * dto.doneFP / dto.consumedHours;
      dto.hoursConsumedForNotDone = sprint.getItems().stream()
          .filter(item -> item.getStatus() != BacklogItem.Status.DONE)
          .mapToInt(BacklogItem::getHoursConsumed).sum();
      if (sprint.getEnd().isAfter(sprint.getPlannedEnd())) {
         dto.delayDays = sprint.getPlannedEnd().until(sprint.getEnd()).getDays();
      }
      return dto;
   }
}

File: ddd-application/src/main/java/victor/training/ddd/agile/service/SprintMetricsService.kt
package victor.training.ddd.agile.service

import org.springframework.stereotype.Service
import org.springframework.web.bind.annotation.PathVariable
import victor.training.ddd.agile.dto.SprintMetrics
import victor.training.ddd.agile.entity.BacklogItem
import victor.training.ddd.agile.entity.Sprint
import victor.training.ddd.agile.entity.SprintItem
import victor.training.ddd.agile.repo.SprintRepo
import java.util.stream.Collectors

@Service
class SprintMetricsService(private val sprintRepo: SprintRepo) {
    fun getSprintMetrics(@PathVariable id: Long): SprintMetrics {
        val sprint = sprintRepo.findOneById(id)
        check(sprint.status() == Sprint.SprintStatus.FINISHED)
        val doneItems = sprint.items().stream()
            .filter { item-> item.status == SprintItem.ItemStatus.DONE }
            .collect(Collectors.toList())
        val dto = SprintMetrics()
        dto.consumedHours = sprint.items().stream().mapToInt { obj-> obj.hoursConsumed() }.sum()
        dto.calendarDays = sprint.startDate()?.until(sprint.endDate())?.days?:0
        dto.doneFP = doneItems.stream().mapToInt { obj-> obj.fpEstimation }.sum()
        dto.fpVelocity = 1.0 * dto.doneFP / dto.consumedHours
        dto.hoursConsumedForNotDone = sprint.items().stream()
            .filter { item-> item.status != SprintItem.ItemStatus.DONE }
            .mapToInt { obj-> obj.hoursConsumed() }.sum()
        if (sprint.endDate()!!.isAfter(sprint.plannedEndDate)) {
            dto.delayDays = sprint.plannedEndDate!!.until(sprint.endDate()).days
        }
        return dto
    }
}